internal/eventtracker/eventtracker.go

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright 2024 The Chainloop Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventtracker
+
+import (
+	"context"
+	"os"
+	"runtime"
+
+	schemaapi "github.com/chainloop-dev/chainloop/app/controlplane/api/workflowcontract/v1"
+	"github.com/chainloop-dev/chainloop/internal/attestation/crafter"
+	"github.com/denisbrodbeck/machineid"
+	"github.com/posthog/posthog-go"
+	"github.com/rs/zerolog"
+)
+
+// Tags represents a collection of event tags.
+type Tags map[string]string
+
+// TrackerClient defines the interface for tracking events.
+type TrackerClient interface {
+	TrackEvent(ctx context.Context, eventName string, tags Tags) error
+}
+
+// CommandTracker is a basic implementation of the Tracker interface.
+type CommandTracker struct {
+	client TrackerClient
+}
+
+// NewCommandTracker creates a new CommandTracker instance.
+func NewCommandTracker(client TrackerClient) *CommandTracker {
+	return &CommandTracker{
+		client: client,
+	}
+}
+
+// NewPosthogTrackerClient creates a new PosthogTracker instance.
+func NewPosthogTrackerClient(apiKey string, serverURL string) *PosthogTracker {
+	client, _ := posthog.NewWithConfig(apiKey, posthog.Config{
+		Endpoint: serverURL,
+	})
+
+	return &PosthogTracker{
+		client: client,
+	}
+}
+
+type NoopTracker struct{}
+
+func NewNoopTrackerClient() *NoopTracker {
+	return &NoopTracker{}
+}
+
+func (n *NoopTracker) TrackEvent(_ context.Context, _ string, _ Tags) error {
+	return nil
+}
+
+// PosthogTracker is an implementation of the TrackerClient interface for PostHog.
+type PosthogTracker struct {
+	client posthog.Client
+}
+
+// TrackEvent sends an event to the PostHog server.
+func (p *PosthogTracker) TrackEvent(_ context.Context, eventName string, tags Tags) error {
+	msg := posthog.Capture{
+		DistinctId: "default-id",
+		Event:      eventName,
+	}
+
+	defaultTags := loadDefaultTags()
+	for k, v := range defaultTags {
+		msg.Properties.Set(k, v)
+	}
+
+	for k, v := range tags {
+		msg.Properties.Set(k, v)
+	}
+
+	machineID, err := machineid.ProtectedID("chainloop")
+	if err == nil {
+		msg.Properties.Set("machine_id", machineID)
+		msg.DistinctId = machineID
+	}
```

Comment:
I'd like to also have a way to record the hashed controlplane URL. I guess that could be done by using the custom tags no?

Reply:
That could be passed as a custom tag right.